### PR TITLE
182548234/client search

### PR DESCRIPTION
**Pull request by ZachCMP**
@awisema @eanders This is ready from the GraphQL side, but I need you all to plug in the client search logic. I've added comment with `TODO <SANDY>` where this needs to be done. You should adjust `search_scope` so that it contains the scope of filtered clients before it's passed into `resolve_clients`. That method handles the ordering. Pagination happens elsewhere in the process, so no need to worry about that piece here. Let me know if you have any questions or need a hand.

### CODE REVIEW

File: drivers/hmis/app/graphql/types/date.rb
Comment (eanders) [inline drivers/hmis/app/graphql/types/date.rb:12]: Do we need to catch empty input values?
```suggestion
     return nil unless input_value.present?

      Date.parse(input_value)
```

File: drivers/hmis/app/graphql/types/date_time.rb
Comment (eanders) [inline drivers/hmis/app/graphql/types/date_time.rb:12]: same here if that's true for date

File: drivers/hmis/app/graphql/types/hmis_schema/client.rb
Comment (eanders) [inline drivers/hmis/app/graphql/types/hmis_schema/client.rb:25]: Is there any easy way to infer big chunks of this from the existing table structure?  Something like:
```
[:first_name, :last_name].each { |f| field f, magic_get_format_from_db, null: magic_get_null_value_from_db }
```

Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/client.rb:25]: If nothing else, it seems like we might be able to populate large chunks of this from `app/models/concerns/hmis/structure/client.rb#hmis_configuration`

File: drivers/hmis/app/graphql/types/hmis_schema/client_sort_option.rb
Comment (eanders) [inline drivers/hmis/app/graphql/types/hmis_schema/client_sort_option.rb:14]: Should e define this as `LAST_NAME_DESC` or does the asc/desc get set somewhere else?

File: drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb
Comment (eanders) [inline drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb:28]: `.order(sort_field => sort_direction)` probably won't hold up in the long run.  We've found we're often sorting on things that aren't on the Client table.  Just a note for the future, but we're probably going to want to pass in a `sort_method` which would look something like `arel_table[:LastName].asc` where `arel_table` is something like `c_t` which is defined in ArelHelper.

**Review by eanders — APPROVED** [review]
This looks good to me.  I'm happy to merge, but had a bunch of random questions.  See those inline.

File: drivers/hmis/app/graphql/types/date.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/types/date.rb:12]: `input_value` should never be null, since this shouldn't be called unless a non-null value was passed to the endpoint; But it's probably wise to add something like that just in case

File: drivers/hmis/app/graphql/types/hmis_schema/client.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/types/hmis_schema/client.rb:25]: Yeah, you could dynamically generate this stuff if you wanted. You'd have to handle the type mapping though, which could be tricky in some cases. I'd opt for an approach where we auto-generate the straightforward stuff that comes directly from columns and then directly add the derived or more complex bits (like `enrollments_field`)

Comment (eanders) [reply drivers/hmis/app/graphql/types/hmis_schema/client.rb:25]: Let's leave it as is for now.  It makes me nervous the number of places we'll need to touch when the spec changes every two years, but I don't want to slow down development with a massive abstraction right now.

File: drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb
Comment (ZachCMP) [reply drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb:28]: Yeah, makes sense. `sort_field` comes from an enum, so we do control what values are allowed for that. Regardless, I agree that this probably isn't going to work forever. When the time comes we can (and should) either expand the sorting logic in this concern or defer to some other abstraction to handle it.

Comment (gigxz) [reply drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb:28]: I'm already hitting that with implementing project sort. We want to sort projects by Organization name and Project name. WIP PR here https://github.com/greenriver/hmis-warehouse/pull/2335 but I'm not sure how to best implement the mapping of Enums to their sort functions. @ZachCMP sounds like you had some ideas there –  can we take a look together?

Comment (ZachCMP) [reply drivers/hmis/app/graphql/types/hmis_schema/has_clients.rb:28]: @gigxz I think we ought not put the sorting logic in the enum. My thought is that we put it elsewhere, maybe in the model, maybe in its own utility class or something. What we could do is enumerate the sorting options there and then generate the GraphQL enum from that. That feels more appropriate to me.

**Pull request merged**
